server/src/main/java/io/crate/execution/engine/join/HashJoinBatchIterator.java

```diff
@@ -314,30 +279,87 @@ private boolean buildBufferAndMatchRight() {
     }
 
     private void addToBuffer(Object[] currentRow, int hash) {
-        Values existingRows = buffer.get(hash);
-        if (existingRows == null) {
-            existingRows = new Values();
-            buffer.put(hash, existingRows);
+        HashGroup hashGroup = buffer.get(hash);
+        if (hashGroup == null) {
+            hashGroup = new HashGroup(emitUnmatchedRows);
+            buffer.put(hash, hashGroup);
         }
-        existingRows.items.add(currentRow);
-        numberOfRowsInBuffer++;
+        hashGroup.add(currentRow);
+        numberOfHashGroupsInBuffer++;
     }
 
     private boolean findMatchingRows() {
-        while (leftMatchingRowsIterator.hasNext()) {
-            leftRow.cells(leftMatchingRowsIterator.next());
+        while (matchedHashGroupRowsIterator.hasNext()) {
+            int currentIdx = matchedHashGroupRowsIterator.nextIndex();
+            leftRow.cells(matchedHashGroupRowsIterator.next());
             combiner.setLeft(leftRow);
             if (joinCondition.test(combiner.currentElement())) {
+                matchedHashGroup.markMatched(currentIdx);
                 return true;
             }
         }
+        matchedHashGroup = null;
+        matchedHashGroupRowsIterator = null;
         return false;
     }
 
-    private static final class Values {
+    private static final class HashGroup implements Iterable<Object[]> {
+
+        private final List<Object[]> rows = new ArrayList<>();
+        @Nullable
+        private final BitSet rowIsJoinedFlags;
```

Comment:
`rows` is  already accounted in the caller (same for old structure`Values` and new structure `HashGroup`).
 But `BitSet` is something new and can take non-trivial amount of memory, it has `long[]` array that can expand on demand. 

I think we need to keep prev length and account for `delta_length * LONG_TYPE_SIZE` if expansion has happened.

I think both `length()` and `size()` in `BitSet` don't fit, we need `words.length` and we can infer it from size like `internalArrayLength = size() / BITS_PER_WORD`

Reply:
Thank you, I observed some leniency in the memory accounting so was not sure to expand the logic to account for  `BitSet` which is relatively small and also considering the cost of accounting.

Could you check https://github.com/crate/crate/pull/18754/commits/a6e774051b48cffc09d62053422c37d73a6a2612 ? This commit adds accounting for `HashGroup` and `BitSet` object headers as well which will take up more space than `BitSet.words` when all rows are mapped to distinct HashGroups for example.

I will run benchmarks again once this part is approved.

Reply:
How do you think of merging this and handling memory accounting separately? I can add a card to CrateDB board so it will be handled before the next release.

Reply:
Let's merge it without BitSet accounting, 5x slow down is not acceptable and 4MB under-accounting is okay for a short allocation